### PR TITLE
mgr/dashboard: add feature toggle for NFS and fix feature toggles regression

### PR DESCRIPTION
**Pull request by epuertat**
Changes:
- New NFS feature toggle added.
- Fixed regression which broke FeatureToggles in the main menu.
- Added extensive unit testing to NavigationComponent
- Added ng-mocks to improve test isolation and performance (~7x, 139s to
20s)
- Added ng-bullet package to improve unit testing performance (2x, 20s
to 9s)
- Added Rxjs Schedulers to implement NgZone.runOutsideAngular in a Rxjs
friendly way (based on
    https://stackoverflow.com/questions/43121400/run-ngrx-effect-outside-of-angulars-zone-to-prevent-timeout-in-protractor)

Minor issues found from exhaustive unit testing:
- Missing permissions in Cluster menu:
  - `permissions.log.read` and `permissions.prometheus.read`
- Missing classes:
  - Block -> Images: `tc_submenuitem_block_images`
  - Block -> iSCSI: `tc_submenuitem_block_iscsi`
- Typos:
  - class `tc_submenuitem_hosts` assigned to OSD menu (instead of
      `tc_submenuitem_osd`)
  - `tc_menuitem_cephs` -> `tc_menuitem_cephfs`

Minor changes:
- Previously, Cluster Map -> CRUSH Map required both OSD and Host
permissions. Now it only requires OSD permissions (there are no
references to hosts in CRUSH Map page). Nevertheless, all system roles
setting OSD permission also set Host's, so no impact expected.
- Previously, Cluster -> Monitoring menu was hidden when both Prometheus
or Alertmanager weren't configured. Now it's displayed and when clicked
on it shows the helper banner indicating that further configuration is 
required. This change removes the dependency on the PrometheusService.

### Before Unit test refactoring [139 secs]
```
 PASS  src/app/core/navigation/navigation/navigation.component.spec.ts (138.036s)
  NavigationComponent
    Test Permissions
      ✓ When disabled permissions: ["hosts"] => hidden: ".tc_submenuitem_hosts,.tc_submenuitem_cluster_inventory,.tc_submenuitem_cluster_services" (3170ms)
      ✓ When disabled permissions: ["monitor"] => hidden: ".tc_submenuitem_cluster_monitor" (2715ms)
      ✓ When disabled permissions: ["osd"] => hidden: ".tc_submenuitem_osds,.tc_submenuitem_crush" (2843ms)
      ✓ When disabled permissions: ["configOpt"] => hidden: ".tc_submenuitem_configuration,.tc_submenuitem_modules" (2973ms)
      ✓ When disabled permissions: ["log"] => hidden: ".tc_submenuitem_log" (2632ms)
      ✓ When disabled permissions: ["prometheus"] => hidden: ".tc_submenuitem_monitoring" (2597ms)
      ✓ When disabled permissions: ["pool"] => hidden: ".tc_menuitem_pool" (2637ms)
      ✓ When disabled permissions: ["rbdImage"] => hidden: ".tc_submenuitem_block_images" (3140ms)
      ✓ When disabled permissions: ["rbdMirroring"] => hidden: ".tc_submenuitem_block_mirroring" (4018ms)
      ✓ When disabled permissions: ["iscsi"] => hidden: ".tc_submenuitem_block_iscsi" (3112ms)
      ✓ When disabled permissions: ["rbdImage","rbdMirroring","iscsi"] => hidden: ".tc_menuitem_block" (3269ms)
      ✓ When disabled permissions: ["nfs"] => hidden: ".tc_menuitem_nfs" (2971ms)
      ✓ When disabled permissions: ["cephfs"] => hidden: ".tc_menuitem_cephfs" (3265ms)
      ✓ When disabled permissions: ["rgw"] => hidden: ".tc_menuitem_rgw,.tc_submenuitem_rgw_daemons,.tc_submenuitem_rgw_buckets,.tc_submenuitem_rgw_users" (3297ms)
      ✓ When enabled permissions: ["hosts"] => visible: ".tc_submenuitem_hosts,.tc_submenuitem_cluster_inventory,.tc_submenuitem_cluster_services" (2995ms)
      ✓ When enabled permissions: ["monitor"] => visible: ".tc_submenuitem_cluster_monitor" (3907ms)
      ✓ When enabled permissions: ["osd"] => visible: ".tc_submenuitem_osds,.tc_submenuitem_crush" (2638ms)
      ✓ When enabled permissions: ["configOpt"] => visible: ".tc_submenuitem_configuration,.tc_submenuitem_modules" (2539ms)
      ✓ When enabled permissions: ["log"] => visible: ".tc_submenuitem_log" (2878ms)
      ✓ When enabled permissions: ["prometheus"] => visible: ".tc_submenuitem_monitoring" (3207ms)
      ✓ When enabled permissions: ["pool"] => visible: ".tc_menuitem_pool" (2619ms)
      ✓ When enabled permissions: ["rbdImage"] => visible: ".tc_submenuitem_block_images" (3097ms)
      ✓ When enabled permissions: ["rbdMirroring"] => visible: ".tc_submenuitem_block_mirroring" (2892ms)
      ✓ When enabled permissions: ["iscsi"] => visible: ".tc_submenuitem_block_iscsi" (4005ms)
      ✓ When enabled permissions: ["rbdImage","rbdMirroring","iscsi"] => visible: ".tc_menuitem_block" (3246ms)
      ✓ When enabled permissions: ["nfs"] => visible: ".tc_menuitem_nfs" (2998ms)
      ✓ When enabled permissions: ["cephfs"] => visible: ".tc_menuitem_cephfs" (3084ms)
      ✓ When enabled permissions: ["rgw"] => visible: ".tc_menuitem_rgw,.tc_submenuitem_rgw_daemons,.tc_submenuitem_rgw_buckets,.tc_submenuitem_rgw_users" (3098ms)
    Test FeatureToggles
      ✓ When disabled features: ["rbd"] => hidden: ".tc_submenuitem_block_images" (2979ms)
      ✓ When disabled features: ["mirroring"] => hidden: ".tc_submenuitem_block_mirroring" (2859ms)
      ✓ When disabled features: ["iscsi"] => hidden: ".tc_submenuitem_block_iscsi" (2652ms)
      ✓ When disabled features: ["rbd","mirroring","iscsi"] => hidden: ".tc_menuitem_block" (3198ms)
      ✓ When disabled features: ["nfs"] => hidden: ".tc_menuitem_nfs" (2930ms)
      ✓ When disabled features: ["cephfs"] => hidden: ".tc_menuitem_cephfs" (4685ms)
      ✓ When disabled features: ["rgw"] => hidden: ".tc_menuitem_rgw,.tc_submenuitem_rgw_daemons,.tc_submenuitem_rgw_buckets,.tc_submenuitem_rgw_users" (3163ms)
      ✓ When enabled features: ["rbd"] => visible: ".tc_submenuitem_block_images" (3305ms)
      ✓ When enabled features: ["mirroring"] => visible: ".tc_submenuitem_block_mirroring" (3276ms)
      ✓ When enabled features: ["iscsi"] => visible: ".tc_submenuitem_block_iscsi" (3200ms)
      ✓ When enabled features: ["rbd","mirroring","iscsi"] => visible: ".tc_menuitem_block" (2990ms)
      ✓ When enabled features: ["nfs"] => visible: ".tc_menuitem_nfs" (3378ms)
      ✓ When enabled features: ["cephfs"] => visible: ".tc_menuitem_cephfs" (3076ms)
      ✓ When enabled features: ["rgw"] => visible: ".tc_menuitem_rgw,.tc_submenuitem_rgw_daemons,.tc_submenuitem_rgw_buckets,.tc_submenuitem_rgw_users" (2858ms)

Test Suites: 1 passed, 1 total
Tests:       42 passed, 42 total
Snapshots:   0 total
Time:        138.995s

```
### After Unit test refactoring [ 10 secs ]
```
  NavigationComponent
    Test Permissions
      ✓ When disabled permissions: ["hosts"] => hidden: ".tc_submenuitem_hosts,.tc_submenuitem_cluster_inventory,.tc_submenuitem_cluster_services" (99ms)
      ✓ When disabled permissions: ["monitor"] => hidden: ".tc_submenuitem_cluster_monitor" (30ms)
      ✓ When disabled permissions: ["osd"] => hidden: ".tc_submenuitem_osds,.tc_submenuitem_crush" (32ms)
      ✓ When disabled permissions: ["configOpt"] => hidden: ".tc_submenuitem_configuration,.tc_submenuitem_modules" (31ms)
      ✓ When disabled permissions: ["log"] => hidden: ".tc_submenuitem_log" (29ms)
      ✓ When disabled permissions: ["prometheus"] => hidden: ".tc_submenuitem_monitoring" (24ms)
      ✓ When disabled permissions: ["pool"] => hidden: ".tc_menuitem_pool" (50ms)
      ✓ When disabled permissions: ["rbdImage"] => hidden: ".tc_submenuitem_block_images" (51ms)
      ✓ When disabled permissions: ["rbdMirroring"] => hidden: ".tc_submenuitem_block_mirroring" (36ms)
      ✓ When disabled permissions: ["iscsi"] => hidden: ".tc_submenuitem_block_iscsi" (40ms)
      ✓ When disabled permissions: ["rbdImage","rbdMirroring","iscsi"] => hidden: ".tc_menuitem_block" (35ms)
      ✓ When disabled permissions: ["nfs"] => hidden: ".tc_menuitem_nfs" (93ms)
      ✓ When disabled permissions: ["cephfs"] => hidden: ".tc_menuitem_cephfs" (49ms)
      ✓ When disabled permissions: ["rgw"] => hidden: ".tc_menuitem_rgw,.tc_submenuitem_rgw_daemons,.tc_submenuitem_rgw_buckets,.tc_submenuitem_rgw_users" (32ms)
      ✓ When enabled permissions: ["hosts"] => visible: ".tc_submenuitem_hosts,.tc_submenuitem_cluster_inventory,.tc_submenuitem_cluster_services" (20ms)
      ✓ When enabled permissions: ["monitor"] => visible: ".tc_submenuitem_cluster_monitor" (12ms)
      ✓ When enabled permissions: ["osd"] => visible: ".tc_submenuitem_osds,.tc_submenuitem_crush" (9ms)
      ✓ When enabled permissions: ["configOpt"] => visible: ".tc_submenuitem_configuration,.tc_submenuitem_modules" (11ms)
      ✓ When enabled permissions: ["log"] => visible: ".tc_submenuitem_log" (11ms)
      ✓ When enabled permissions: ["prometheus"] => visible: ".tc_submenuitem_monitoring" (9ms)
      ✓ When enabled permissions: ["pool"] => visible: ".tc_menuitem_pool" (9ms)
      ✓ When enabled permissions: ["rbdImage"] => visible: ".tc_submenuitem_block_images" (8ms)
      ✓ When enabled permissions: ["rbdMirroring"] => visible: ".tc_submenuitem_block_mirroring" (9ms)
      ✓ When enabled permissions: ["iscsi"] => visible: ".tc_submenuitem_block_iscsi" (7ms)
      ✓ When enabled permissions: ["rbdImage","rbdMirroring","iscsi"] => visible: ".tc_menuitem_block" (9ms)
      ✓ When enabled permissions: ["nfs"] => visible: ".tc_menuitem_nfs" (6ms)
      ✓ When enabled permissions: ["cephfs"] => visible: ".tc_menuitem_cephfs" (10ms)
      ✓ When enabled permissions: ["rgw"] => visible: ".tc_menuitem_rgw,.tc_submenuitem_rgw_daemons,.tc_submenuitem_rgw_buckets,.tc_submenuitem_rgw_users" (9ms)
    Test FeatureToggles
      ✓ When disabled features: ["rbd"] => hidden: ".tc_submenuitem_block_images" (11ms)
      ✓ When disabled features: ["mirroring"] => hidden: ".tc_submenuitem_block_mirroring" (13ms)
      ✓ When disabled features: ["iscsi"] => hidden: ".tc_submenuitem_block_iscsi" (18ms)
      ✓ When disabled features: ["rbd","mirroring","iscsi"] => hidden: ".tc_menuitem_block" (11ms)
      ✓ When disabled features: ["nfs"] => hidden: ".tc_menuitem_nfs" (17ms)
      ✓ When disabled features: ["cephfs"] => hidden: ".tc_menuitem_cephfs" (13ms)
      ✓ When disabled features: ["rgw"] => hidden: ".tc_menuitem_rgw,.tc_submenuitem_rgw_daemons,.tc_submenuitem_rgw_buckets,.tc_submenuitem_rgw_users" (12ms)
      ✓ When enabled features: ["rbd"] => visible: ".tc_submenuitem_block_images" (12ms)
      ✓ When enabled features: ["mirroring"] => visible: ".tc_submenuitem_block_mirroring" (10ms)
      ✓ When enabled features: ["iscsi"] => visible: ".tc_submenuitem_block_iscsi" (10ms)
      ✓ When enabled features: ["rbd","mirroring","iscsi"] => visible: ".tc_menuitem_block" (15ms)
      ✓ When enabled features: ["nfs"] => visible: ".tc_menuitem_nfs" (10ms)
      ✓ When enabled features: ["cephfs"] => visible: ".tc_menuitem_cephfs" (11ms)
      ✓ When enabled features: ["rgw"] => visible: ".tc_menuitem_rgw,.tc_submenuitem_rgw_daemons,.tc_submenuitem_rgw_buckets,.tc_submenuitem_rgw_users" (10ms)

Test Suites: 1 passed, 1 total
Tests:       42 passed, 42 total
Snapshots:   0 total
Time:        9.609s

```
Fixes: https://tracker.ceph.com/issues/43419
Fixes: https://tracker.ceph.com/issues/43715
Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

## Checklist
- [x] References tracker ticket
- [x] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by callithea** [comment]
jenkins test make check

**Comment by callithea** [comment]
jenkins test dashboard backend

File: src/pybind/mgr/dashboard/frontend/src/app/core/navigation/navigation/navigation.component.html [hunk 1]
```diff
@@ -210,7 +210,7 @@
   <!-- NFS -->
   <li routerLinkActive="active"
       class="item tc_menuitem_nfs"
-      *ngIf="permissions?.nfs?.read">
+      *ngIf="enabledFeature.nfs && permissions?.nfs?.read">
```
Comment (votdev) [inline src/pybind/mgr/dashboard/frontend/src/app/core/navigation/navigation/navigation.component.html]:
```suggestion
      *ngIf="enabledFeature$.nfs && permissions?.nfs?.read">
```

Comment (epuertat) [reply src/pybind/mgr/dashboard/frontend/src/app/core/navigation/navigation/navigation.component.html]: Thanks for the suggestion @votdev! The error actually is caused by the fact that feature-toggles code in Navigation component was (accidentally) removed during Boostrap upgrade (#25188). I'll need to fix that too, but given @tspmelo is working on the Navigation component right now (#31923), I'll defer this work until that's merged.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/directives/feature-toggles.directive.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import { Subscription } from 'rxjs';
+import { distinctUntilChanged, pluck } from 'rxjs/operators';
+
+import { FeatureTogglesService } from '../services/feature-toggles.service';
+
+
+@Directive({
+  selector: '[cdFeature]',
+})
+export class FeatureToggleDirective implements OnInit, OnDestroy {
```
[the remark this answers is not in the document]
Comment (epuertat) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/directives/feature-toggles.directive.ts]: Thanks @Devp00l ! I've removed that new directive (moved to the [NavigationComponent refactor](https://github.com/ceph/ceph/pull/33375)) and took a simpler approach.

**Comment by epuertat** [comment]
jenkins test make check

**Comment by epuertat** [comment]
jenkins test dashboard

**Comment by alfonsomthd** [comment]
jenkins test make check

File: src/pybind/mgr/dashboard/frontend/src/app/core/navigation/navigation/navigation.component.ts [hunk 3]
```diff
@@ -19,12 +18,12 @@ export class NavigationComponent implements OnInit {
   @HostBinding('class.isPwdDisplayed') isPwdDisplayed = false;
 
   permissions: Permissions;
+  enabledFeature$: FeatureTogglesMap$;
   summaryData: any;
   icons = Icons;
 
   isAlertmanagerConfigured = false;
```
Comment (alfonsomthd) [inline src/pybind/mgr/dashboard/frontend/src/app/core/navigation/navigation/navigation.component.ts]: Please remove these 2 unused variables.

**Comment by alfonsomthd** [comment]
jenkins test dashboard backend

**Comment by votdev** [comment]
jenkins test make check

**Comment by callithea** [comment]
- [pkg build](https://shaman.ceph.com/builds/ceph/wip-laura-testing-32419/) [WIP]

**Comment by epuertat** [comment]
> I'm not able to hide any feature in the UI via `ceph dashboard feature disable iscsi mirroring nfs`. 

@votdev weird... I just tried once more and _it works for me_. It takes a while, though, as I increased polling period from 20 secs to 30 secs, and also takes a few extra seconds to update. What's your setup? Mine is Chrome-based.

**Comment by votdev** [comment]
> > I'm not able to hide any feature in the UI via `ceph dashboard feature disable iscsi mirroring nfs`.
> 
> @votdev weird... I just tried once more and _it works for me_. It takes a while, though, as I increased polling period from 20 secs to 30 secs, and also takes a few extra seconds to update. What's your setup? Mine is Chrome-based.

I am using Firefox. And i waited more than the timeout period ;-)

**Comment by alfonsomthd** [comment]
> > > I'm not able to hide any feature in the UI via `ceph dashboard feature disable iscsi mirroring nfs`.
> > 
> > 
> > @votdev weird... I just tried once more and _it works for me_. It takes a while, though, as I increased polling period from 20 secs to 30 secs, and also takes a few extra seconds to update. What's your setup? Mine is Chrome-based.
> 
> I am using Firefox. And i waited more than the timeout period ;-)

@votdev Strange becasue this morning I tested it on Firefox and it worked as expected.
It takes some seconds to apply.

**Comment by epuertat** [comment]
@votdev just tried in Firefox (72.0.2 (64-bit)) and it works ok (it refreshes even faster than Chrome).

**Comment by votdev** [comment]
> @votdev just tried in Firefox (72.0.2 (64-bit)) and it works ok (it refreshes even faster than Chrome).

I'm sorry, it seems i had some troubles with my vstart environment. After restarting the Docker and vstart, i was able to successfully hide the features in the UI.

The only thing i would like to see improved is that the timer in the `FeatureTogglesService`should immediatelly fire on start and not after xxx seconds.

**Comment by epuertat** [comment]
After investigating with @votdev it all seemed to be caused by environment issues. Nevertheless, just refreshed this PR changing `rxjs` operator `subscribeOn` with `observeOn`, as it speeds up UI updates after HTTP response.

**Comment by epuertat** [comment]
jenkins test dashboard backend

**Pull request merged**